### ISSUE TITLE
Could System.Xml.XmlNamespaceManager be made fluent?

### ISSUE THREAD

**Issue by doomchild** [dotnet/corefx#36666]
So this is a pretty small thing that I think could be done in several places throughout the framework, but I think it's useful for a couple of different reasons.  I picked System.Xml.XmlNamespaceManager just because it seemed like a very localized change.

Specifically, I'm thinking about code that at the moment has to be written like this:

```csharp
private static XmlNamespaceManager CreateNamespaceManager()
{
    var nsManager = new XmlNamespaceManager(new NameTable());
    
    nsManager.AddNamespace(prefix1, ns1);
    nsManager.AddNamespace(prefix2, ns2);
    nsManager.AddNamespace(prefix3, ns3);
    nsManager.AddNamespace(prefix4, ns4);
    nsManager.AddNamespace(prefix5, ns5);
    nsManager.AddNamespace(prefix6, ns6);
    nsManager.AddNamespace(prefix7, ns7);

    return nsManager;
}
```

I would much rather write it like this:

```csharp
private static XmlNamespaceManager CreateNamespaceManager()
{
    return new XmlNamespaceManager(new NameTable())
	.AddNamespace(prefix1, ns1)
	.AddNamespace(prefix2, ns2)
	.AddNamespace(prefix3, ns3)
	.AddNamespace(prefix4, ns4)
	.AddNamespace(prefix5, ns5)
	.AddNamespace(prefix6, ns6)
	.AddNamespace(prefix7, ns7);
}
```

For one, it's less to type.  I get that that's not a super compelling reason, but it's a nice to have.  What this allows you to do, however, is support functional workflows like this (using a `Maybe` that we've built in-house, you can just imagine using `language-ext` if you want):

```csharp
var nodes = Maybe.OfNullable(xmlDocument)
    .Map(xmlDoc => xmlDoc.NameTable)
    .Map(nameTable => new XmlNamespaceManager(nameTable))
    .Map(xmlNsMgr => xmlNsMgr.RemoveNamespace(docPrefix1, docNs1))
    .Map(xmlNsMgr => xmlNsMgr.AddNamespace(prefix1, ns1))
    .Map(xmlNsMgr => xmlNsMgr.AddNamespace(prefix2, ns2))
    .Map(xmlNsMgr => xmlDocument.SelectNodes(someXPath, xmlNsMgr))
    .GetOrDefault(new ArrayList());
```

Is this a reasonable idea?

**Comment by danmosemsft** [comment]
We've generally avoided fluent style in the base libraries for various reasons. Could you achieve this with a simple wrapper class?

**Comment by svick** [comment]
Do I understand it correctly that your workflow is the equivalent of the following code?

```c#
XmlNodeList nodes;

if (xmlDocument != null)
{
    var xmlNsMgr = new XmlNamespaceManager(xmlDocument.NameTable);
    xmlNsMgr.RemoveNamespace(docPrefix1, docNs1);
    xmlNsMgr.AddNamespace(prefix1, ns1);
    xmlNsMgr.AddNamespace(prefix2, ns2);
    nodes = xmlDocument.SelectNodes(someXPath, xmlNsMgr);
}

nodes = nodes ?? new ArrayList();
```

If that's the case, then I'm not sure it's a good argument for this change, especially since most people are not using those kinds of workflows.

**Comment by doomchild** [comment]
@svick: No, that was a wholly contrived example.  I'm a little leery of just copying our code and putting it up on Github.

@danmosemsft:  Is there a reason that fluency is avoided in the base libraries?  It seems to me that you can write the exact same code you can now if the methods are fluent as you can if they aren't.  The only change is that you can _also_ chain them more readably if you like.

**Comment by danmosemsft** [comment]
@terrajobst can answer

**Comment by terrajobst** [comment]
This would feel very over enginneered. We could, however, consider adding a method called `Add` (as opposed to `AddNamespace`. This would allow you to use collection initializer syntax, like so:

```C#
private static XmlNamespaceManager CreateNamespaceManager()
    => new XmlNamespaceManager(new NameTable())
       {
           {prefix1, ns1 }
           {prefix2, ns2 }
           {prefix3, ns3 }
           {prefix4, ns4 }
           {prefix5, ns5 }
           {prefix6, ns6 }
           {prefix7, ns7 }
       };
```

Fluent makes sense when you have a builder like pattern and need to support vastly different operations. For things like this it feels heavy handed.

**Comment by doomchild** [comment]
I guess the reason I suggested altering the existing methods to return fluently (instead of adding a new method to the type) is that I didn't want to add a new method if changing an existing one could fit existing and new use cases instead.  I'm certainly not opposed to an `Add` method, though.

**Comment by svick** [comment]
@doomchild 

> I didn't want to add a new method if changing an existing one could fit existing and new use cases instead

Changing the return type of an existing method is a [binary breaking change](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#property-field-parameter-and-return-values). So adding a new method is actually much more likely to get accepted.
